File: golox/internal/scanner/scanner.go
package scanner

import (
	"fmt"
	"strconv"
	"unicode/utf8"

	"github.com/michael-go/lox/golox/internal/globals"
	"github.com/michael-go/lox/golox/internal/token"
)

type Scanner struct {
	source string
	tokens []token.Token

	start   int
	current int
	line    int
}

func New(source string) Scanner {
	s := Scanner{source: source, line: 1}
	return s
}

func (s *Scanner) ScanTokens() ([]token.Token, error) {
	for !s.isAtEnd() {
		s.start = s.current
		s.scanToken()
	}

	s.start = s.current
	s.addToken(token.EOF)

	return s.tokens, nil
}

func (s *Scanner) isAtEnd() bool {
	return s.current >= len(s.source)
}

func (s *Scanner) scanToken() {
	r := s.advance()
	switch r {
	case rune('('):
		s.addToken(token.LEFT_PAREN)
	case rune(')'):
		s.addToken(token.RIGHT_PAREN)
	case rune('{'):
		s.addToken(token.LEFT_BRACE)
	case rune('}'):
		s.addToken(token.RIGHT_BRACE)
	case rune(','):
		s.addToken(token.COMMA)
	case rune('.'):
		s.addToken(token.DOT)
	case rune('-'):
		s.addToken(token.MINUS)
	case rune('+'):
		s.addToken(token.PLUS)
	case rune(';'):
		s.addToken(token.SEMICOLON)
	case rune('*'):
		s.addToken(token.STAR)
	case rune('!'):
		if s.match('=') {
			s.addToken(token.BANG_EQUAL)
		} else {
			s.addToken(token.BANG)
		}
	case rune('='):
		if s.match('=') {
			s.addToken(token.EQUAL_EQUAL)
		} else {
			s.addToken(token.EQUAL)
		}
	case rune('<'):
		if s.match('=') {
			s.addToken(token.LESS_EQUAL)
		} else {
			s.addToken(token.LESS)
		}
	case rune('>'):
		if s.match('=') {
			s.addToken(token.GREATER_EQUAL)
		} else {
			s.addToken(token.GREATER)
		}
	case rune('/'):
		if s.match('/') {
			for !s.isAtEnd() && s.peek() != '\n' {
				s.advance()
			}
		} else {
			s.addToken(token.SLASH)
		}
	case rune(' '):
	case rune('\r'):
	case rune('\t'):
	case rune('\n'):
		s.line++
	case rune('"'):
		s.string()
	default:
		if isDigit(r) {
			s.number()
		} else if isAlpha(r) {
			s.identifier()
		} else {
			globals.ReportError(s.line, "", fmt.Sprintf("Unexpected character %#U", r))
		}
	}
}

func (s *Scanner) advance() rune {
	r, len := utf8.DecodeRuneInString(s.source[s.current:])
	s.current += len
	return r
}

func (s *Scanner) addToken(t token.Type) {
	s.addTokenLiteral(t, nil)
}

func (s *Scanner) addTokenLiteral(tokenType token.Type, literal any) {
	text := s.source[s.start:s.current]
	s.tokens = append(s.tokens, token.Token{Type: tokenType, Lexeme: text, Literal: literal, Line: s.line})
}

func (s *Scanner) match(expected rune) bool {
	if s.isAtEnd() {
		return false
	}

	r, _ := utf8.DecodeRuneInString(s.source[s.current:])
	if r != expected {
		return false
	}

	s.advance()
	return true
}

func (s *Scanner) peek() rune {
	if s.isAtEnd() {
		return rune(0)
	}

	r, _ := utf8.DecodeRuneInString(s.source[s.current:])
	return r
}

func (s *Scanner) peekNext() rune {
	if s.current+1 >= len(s.source) {
		return rune(0)
	}

	r, _ := utf8.DecodeRuneInString(s.source[s.current+1:])
	return r
}

func (s *Scanner) string() {
	for !s.isAtEnd() && s.peek() != '"' {
		if s.peek() == '\n' {
			s.line++
		}
		s.advance()
	}

	if s.isAtEnd() {
		globals.ReportError(s.line, "", "Unterminated string.")
		return
	}

	s.advance()

	value := s.source[s.start+1 : s.current-1]
	s.addTokenLiteral(token.STRING, value)
}

func (s *Scanner) number() {
	for isDigit(s.peek()) {
		s.advance()
	}

	if s.peek() == '.' && isDigit(s.peekNext()) {
		s.advance()
		for isDigit(s.peek()) {
			s.advance()
		}
	}

	value, _ := strconv.ParseFloat(s.source[s.start:s.current], 64)
	s.addTokenLiteral(token.NUMBER, value)
}

func (s *Scanner) identifier() {
	for isAlphaNumeric(s.peek()) {
		s.advance()
	}

	text := s.source[s.start:s.current]
	tokenType, exists := keywords[text]
	if !exists {
		tokenType = token.IDENTIFIER
	}
	s.addToken(tokenType)
}

File: rslox/src/scanner.rs
#[derive(PartialEq, Debug, Clone, Copy)]
pub enum TokenKind {
    // Single character tokens:
    LeftParen = 0,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens:
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // Literals:
    Identifier,
    String,
    Number,
    // Keywords:
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
    // Error & EOF:
    Error,
    Eof,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String, // TODO: try to make it a ref
    pub line: usize,
}

pub struct Scanner {
    source_chars: Vec<char>,
    start_pos: usize,
    current_pos: usize,
    line: usize,
}

impl Scanner {
    pub fn new(source: &str) -> Scanner {
        Scanner {
            // this simplifies random access to the utf8 string,
            //  but we can make it more efficient by holding the ref to original string instead,
            //  and using start_pos and current_pos to index the utf8 byte positions
            source_chars: source.chars().collect(),
            start_pos: 0,
            current_pos: 0,
            line: 1,
        }
    }

    pub fn scan_token(&mut self) -> Token {
        self.skip_whitespace();
        self.start_pos = self.current_pos;

        if self.is_at_end() {
            return self.make_token(TokenKind::Eof);
        }

        let c = self.advance();

        if is_digit(c) {
            return self.number();
        }
        if is_alpha(c) {
            return self.identifier();
        }

        match c {
            '(' => self.make_token(TokenKind::LeftParen),
            ')' => self.make_token(TokenKind::RightParen),
            '{' => self.make_token(TokenKind::LeftBrace),
            '}' => self.make_token(TokenKind::RightBrace),
            ',' => self.make_token(TokenKind::Comma),
            '.' => self.make_token(TokenKind::Dot),
            '-' => self.make_token(TokenKind::Minus),
            '+' => self.make_token(TokenKind::Plus),
            '/' => self.make_token(TokenKind::Slash),
            ';' => self.make_token(TokenKind::Semicolon),
            '*' => self.make_token(TokenKind::Star),
            '!' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::BangEqual)
                } else {
                    self.make_token(TokenKind::Bang)
                }
            }
            '=' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::EqualEqual)
                } else {
                    self.make_token(TokenKind::Equal)
                }
            }
            '<' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::LessEqual)
                } else {
                    self.make_token(TokenKind::Less)
                }
            }
            '>' => {
                if self.match_char('=') {
                    self.make_token(TokenKind::GreaterEqual)
                } else {
                    self.make_token(TokenKind::Greater)
                }
            }
            '"' => self.string(),

            _ => self.error_token("Unexpected character."),
        }
    }

    fn is_at_end(&self) -> bool {
        self.current_pos >= self.source_chars.len()
    }

    fn make_token(&self, kind: TokenKind) -> Token {
        Token {
            kind: kind,
            lexeme: self.source_chars[self.start_pos..self.current_pos]
                .iter()
                .collect(),
            line: self.line,
        }
    }

    fn error_token(&self, message: &str) -> Token {
        Token {
            kind: TokenKind::Error,
            lexeme: message.to_string(),
            line: self.line,
        }
    }

    fn advance(&mut self) -> char {
        let c = self.peek();
        self.current_pos += 1;
        c
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }

        let c = self.peek();
        if c != expected {
            return false;
        }

        self.current_pos += 1;
        true
    }

    fn skip_whitespace(&mut self) {
        loop {
            if self.is_at_end() {
                break;
            }
            let c = self.peek();
            match c {
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }
                '/' => {
                    if self.peek_next() == '/' {
                        // A comment goes until the end of the line.
                        while self.peek() != '\n' && !self.is_at_end() {
                            self.advance();
                        }
                    } else {
                        return;
                    }
                }
                _ => return,
            };
        }
    }

    fn peek(&self) -> char {
        if self.current_pos >= self.source_chars.len() {
            return '\0';
        }
        self.source_chars[self.current_pos]
    }

    fn peek_next(&self) -> char {
        if self.current_pos + 1 >= self.source_chars.len() {
            return '\0';
        }

        self.source_chars[self.current_pos + 1]
    }

    fn string(&mut self) -> Token {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            return self.error_token("Unterminated string.");
        }

        // The closing ".
        self.advance();
        self.make_token(TokenKind::String)
    }

    fn number(&mut self) -> Token {
        while is_digit(self.peek()) {
            self.advance();
        }

        // Look for a fractional part.
        if self.peek() == '.' && is_digit(self.peek_next()) {
            // Consume the "."
            self.advance();

            while is_digit(self.peek()) {
                self.advance();
            }
        }

        self.make_token(TokenKind::Number)
    }

    fn identifier(&mut self) -> Token {
        while is_alpha_numeric(self.peek()) {
            self.advance();
        }

        self.make_token(self.identifier_type())
    }

    fn identifier_type(&self) -> TokenKind {
        match self.source_chars[self.start_pos..self.current_pos]
            .iter()
            .collect::<String>()
            .as_str()
        {
            "and" => TokenKind::And,
            "class" => TokenKind::Class,
            "else" => TokenKind::Else,
            "false" => TokenKind::False,
            "for" => TokenKind::For,
            "fun" => TokenKind::Fun,
            "if" => TokenKind::If,
            "nil" => TokenKind::Nil,
            "or" => TokenKind::Or,
            "print" => TokenKind::Print,
            "return" => TokenKind::Return,
            "super" => TokenKind::Super,
            "this" => TokenKind::This,
            "true" => TokenKind::True,
            "var" => TokenKind::Var,
            "while" => TokenKind::While,
            _ => TokenKind::Identifier,
        }
    }
}

fn is_digit(c: char) -> bool {
    c >= '0' && c <= '9'
}

fn is_alpha(c: char) -> bool {
    (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

fn is_alpha_numeric(c: char) -> bool {
    is_alpha(c) || is_digit(c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn scan() {
        let mut scanner = Scanner::new(
            "
            fun foo() { // comment 
                return 3 * (4 + 5);
            }",
        );
        let mut tokens: Vec<super::Token> = Vec::new();
        loop {
            let token = scanner.scan_token();
            if token.kind == super::TokenKind::Eof {
                break;
            }
            tokens.push(token);
        }

        assert_eq!(
            tokens,
            vec![
                Token {
                    kind: TokenKind::Fun,
                    lexeme: "fun".to_string(),
                    line: 2
                },
                Token {
                    kind: TokenKind::Identifier,
                    lexeme: "foo".to_string(),
                    line: 2
                },
                Token {
                    kind: TokenKind::LeftParen,
                    lexeme: "(".to_string(),
                    line: 2
                },
                Token {
                    kind: TokenKind::RightParen,
                    lexeme: ")".to_string(),
                    line: 2
                },
                Token {
                    kind: TokenKind::LeftBrace,
                    lexeme: "{".to_string(),
                    line: 2
                },
                Token {
                    kind: TokenKind::Return,
                    lexeme: "return".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::Number,
                    lexeme: "3".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::Star,
                    lexeme: "*".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::LeftParen,
                    lexeme: "(".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::Number,
                    lexeme: "4".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::Plus,
                    lexeme: "+".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::Number,
                    lexeme: "5".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::RightParen,
                    lexeme: ")".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::Semicolon,
                    lexeme: ";".to_string(),
                    line: 3
                },
                Token {
                    kind: TokenKind::RightBrace,
                    lexeme: "}".to_string(),
                    line: 4
                },
            ]
        );
    }
}
